Feature/blog observables (#484)
ordina-jworks/ordina-jworks.github.io: _posts/2019-02-05-observables.md

---
layout: post
authors: [tim_vierbergen]
title: 'Observables: The right way'
image: /img/2018-11-21-sse-spring-node-dev-ci/sse-front.png
tags: [Angular,rxjs,Observables,Nodejs]
category: Frontend
comments: true
---

# Table of contents
1. [Intro](#intro)
2. [Setup](#setup)
3. [Refactorin](#refactoring)
4. [Example](#example)

> # Intro #

Reviewer: What's the hashtag for at the end?
```suggestion
# Intro
```

During my consultancy projects, I often come across the same implementations and problems when colleagues are trying to implement an observable strategy.
A lot of frameworks are offering observables out of the box for their communication layer.

> Almost all `Promises` are replace by `Observables` nowadays.

Reviewer 2: typo: "replaced"

It needs to display the number of unread or critical messages.

datatable.component.ts:
```javascript
this._dataService.get(this.filter).subscribe((page:Page) => {
	this.data = page.data;
})
```

header.component.ts:
```javascript
this._dataService.get(this.filter).subscribe((page:Page) => {
	this.count = page.number;
});
```
data.service.ts:
```javascript
public get(filter?: Filter): Observable<Page> {
	return this._http.get('urlToData' + this.parseFilterToString(filter)).pipe(
		tap((data) => this.logService.log(data)),
		map((response) => {
			let page: Page = new Page();
			page.data = response.body;
			page.number = response.headers.get('x-count');
		})
	);
}
```

In the setup we have so far, both components will use the same service for requesting the data.
They will both subscribe to an `Observable`, however, it will be a different one.
When you refresh the data in the datatable, the header will not receive a new value in its subscription and therefore will still show the old number of messages.

In some use cases this might be the desired outcome, but in most cases you want more components to be able to subscribe to the same `Observable`.

> # Setup #

Reviewer: ```suggestion
# Setup
```

In our new setup, we want both, header and datatable, to subscribe to the same Observable, so a call for new data will result in an update in the datatable and the header.
To make this happen, we will use some kind of layer in between.
This new layer will provide our components with one and only one and the same `Observable` and will mask the the communication layer from the view's controller.
Both components will subscribe to this service, so they both get updated by the same result.
We can do this by creating a simple `Subject` in our service and return that Subject as an Observable to our components.
We can then implement other calls for this service that will trigger an update of the data, and send it through this subject to both components.
Because we are not providing a filter when we call the getter for the `Observable` (Subject) we should also find a way of providing the filter to the service, before requerying the data.

> This means we can/are going to use one shared filter, for both components, which makes sense in this case, but not in all use cases.

Reviewer: ```suggestion
This means we are going to use one shared filter for both components, which makes sense in this case, but not in all use cases.
```

<div style="text-align: center;" >

>   <img src="/img/observables-the-right-way/second-step.png" width="60%">

Reviewer: Can also make this one a bit bigger imo.
```suggestion
  <img src="/img/observables-the-right-way/second-step.png" width="85%">
```

</div>

> # Refactoring #

Reviewer: ```suggestion
# Refactoring
```

We actually don't need to refactor any of our components.

> They will still subscribe to an `Observable` of the service, and react on the incomming data.

Reviewer: ```suggestion
They will still subscribe to an `Observable` of the service, and react on the incoming data.
```

The service however will get refactored.

> Start by defining the Subject as a local property, and because we are going to implement a getter, we can make it private.

Reviewer: ```suggestion
Start by defining the `Subject` as a local property, and because we are going to implement a getter, we can make it private.
```

> data.service.ts

Reviewer: ```suggestion
`data.service.ts`
```

```javascript

private _data$: Subject<Page> = new BehavioralSubject<>({});
...

public data(): Observable<Page> {
	return this._data$ as Observable<Page>;
}

...

public reload(): void {
	this._http.get('urlToData' + this.parseFilterToString(this._filter)).pipe(
		tap((data) => this.logService.log(data)),
		map((response) => {
			let page: Page = new Page();
			page.data = response.body;
			page.number = response.headers.get('x-count');
		}).subscribe((page: Page) => {
			this._data$.next(page);
		})

> 	); // and maybe unsubscribe or throw it way or make a Promise out of it

Reviewer: ```suggestion
	); // and maybe unsubscribe, throw it away or make a Promise out of it
```

Author: Removed it... HTTPClient is unsubscribing itself

}

```

> This way every component or service, that is subscribing on this subject, is getting data when some other component or service triggers the `reload`.

Reviewer: ```suggestion
This way, every component or service that is subscribing on this subject, is getting data when some other component or service triggers the `reload`.
```

There are even a lot more options to this setup:
* Clearing data

> * Resetting a default filter

Reviewer: ```suggestion
* Resetting to a default filter
```

> * Refreshing current filter

Reviewer: ```suggestion
* Refreshing the current filter
```

> * Caching data.

Reviewer: ```suggestion
* Caching data
```

* Manipulating data through other services or components
* Adding an event consumer that also updates the datatable
* ...

> # Example #

Reviewer: ```suggestion
# Example
```

> # Example #

Reviewer: You should add a screenshot of the example's frontend.

I've build a simple example to demonstrate this behaviour.
A header that is displaying an alert icon when there are unread messages that are also critical.

> A sidebar that is displaying the amount of unread messages next to it's navigation link, and an overview of the messages, with a basic paging implementation.

Reviewer: ```suggestion
A sidebar that is displaying the amount of unread messages next to its navigation link, and an overview of the messages, with a basic paging implementation.
```

> A simple backend that is written in nodejs with express will provide a few endpoint:

Reviewer: ```suggestion
A simple backend that is written in Node.js with Express will provide a few endpoint:
```

> * api/message // with paging and filter, although the filter isn't implemented in the frontend example.

Reviewer: ```suggestion
* `api/message`: with paging and a filter, although the filter isn't implemented in the frontend example
```

> * api/message/:id // not used

Reviewer: ```suggestion
* `api/message/:id`: not used
```

> * api/stream

Reviewer: ```suggestion
* `api/stream`
```

> * api/refresh

Reviewer: ```suggestion
* `api/refresh`
```

> The service is not reloading data as long as the `page` or the `filter` is changed.

Reviewer: Is changed? or hasn't changed?
```suggestion
The service is not reloading data as long as the `page` or the `filter` hasn't changed.
```

> While a the service is still loading the data, a new reload will not fetch again the data.

Reviewer: ```suggestion
While the service is still loading the data, a new reload will not cause the data to be fetched again.
```

> You can find the code <a target="_blank" href="https://gitlab.com/VeeTeeDev/observables-demo">here</a>

Reviewer: ```suggestion
You can find the code on <a target="_blank" href="https://gitlab.com/VeeTeeDev/observables-demo">GitLab</a>.
```

Server-Sent Events are added to update the `read` status of a message when the envelope gets clicked.
This will also trigger the observable.

> To run front and backend together

Reviewer: ```suggestion
To run front- and backend together, execute the following command in the root of the project:
```

`$ npm run start`

> The way, a proxy is added to the `serve` command to overcome `CORS` blocking going from localhost:4200 to localhost:3000

Reviewer: Don't understand this sentence?

Author: 'This' way...

> Don't mind the backend server, it's a quick and dirty solution and is not implemented as it should.

Reviewer: Is this worth mentioning?

---
layout: post
authors: [tim_vierbergen]
title: 'Observables: The right way'
image: /img/observables-the-right-way/cover.png
tags: [Angular,rxjs,Observables,Nodejs]
category: Frontend
comments: true
---

# Table of contents
1. [Intro](#intro)
2. [Setup](#setup)
3. [Refactoring](#refactoring)
4. [Example](#example)

> 

Reviewer 3: Adding a conclusion would be nice, as the blog has kind of an abrupt ending.

Both components will subscribe to this service, so they both get updated by the same result.
We can do this by creating a simple `Subject` in our service and return that Subject as an Observable to our components.
We can then implement other calls for this service that will trigger an update of the data, and send it through this subject to both components.
Because we are not providing a filter when we call the getter for the `Observable` (Subject) we should also find a way of providing the filter to the service, before requerying the data.
This means we can/are going to use one shared filter, for both components, which makes sense in this case, but not in all use cases.

<div style="text-align: center;" >
  <img src="/img/observables-the-right-way/second-step.png" width="60%">
</div>

# Refactoring #

We actually don't need to refactor any of our components.
They will still subscribe to an `Observable` of the service, and react on the incomming data.
The service however will get refactored.
Start by defining the Subject as a local property, and because we are going to implement a getter, we can make it private.

data.service.ts
```javascript

> private _data$: Subject<Page> = new BehavioralSubject<>({});

Reviewer 3: ```suggestion
private _data$: Subject<Page> = new BehaviorSubject<>({});
```
